cf-testd/cf-testd.c

```diff
@@ -469,12 +542,49 @@ int main(int argc, char *argv[])
     if (config->file != NULL)
     {
         Log(LOG_LEVEL_NOTICE, "Got file argument: '%s'", config->file);
-        Log(LOG_LEVEL_ERR, "File input not yet supported");
-        // TODO: implement file input
-        exit(EXIT_FAILURE);
+        FILE *fp = NULL;
+        char buffer[CF_BUFSIZE];
+        int c, i = 0;
+
+        if ((fp = fopen(config->file, "r")) == NULL)
+        {
+            Log(LOG_LEVEL_INFO,
+                "Could not read report file at %s (fopen: %s)",
+                config->file,
+                GetErrorStr());
+            exit(EXIT_FAILURE);
+        }
+
+        while ((c = getc(fp)) != EOF)
```

Comment:
Sorry for the nitpicking, but is there any reason why the file is read character after character?

Reply:
Nitpicking is good :)

Reply:
Agreed. I don't mind nitpicking at all. No particular reason, what would you suggest?

Reply:
@vpodzime probably just use read() then eh? I'll rewrite it that way.

Reply:
Looking at https://wiki.sei.cmu.edu/confluence/display/c/FIO19-C.+Do+not+use+fseek%28%29+and+ftell%28%29+to+compute+the+size+of+a+regular+file and other bits online, maybe I should use fstat and dynamically allocate enough space for a reasonably sized file... say 1MB? I looked at a sample previous_state dir and got 260K. Certainly much bigger than the previous CF_BUFSIZE of 4K which certainly would have failed us in more strenuous/realistic testing.

Reply:
Oh but wait... it only works with regular files to do any sort of fstat or fseek to determine the size and I want/need to support device files like /dev/stdin. So... @vpodzime let me know what you would suggest.

Reply:
maybe use our existing helper functions:
```
    if (config->file != NULL)
    {
        Log(LOG_LEVEL_NOTICE, "Got file argument: '%s'", config->file);
        if (!FileCanOpen(config->file, "r"))
        {
            Log(LOG_LEVEL_ERR, "Can't open file '%s' for reading", config->file);
            exit(EXIT_FAILURE);
        }

        Writer *contents = FileRead(config->file, SIZE_MAX, NULL);
        if (!contents)
        {
            Log(LOG_LEVEL_ERR, "Error reading report file '%s'", config->file);
            exit(EXIT_FAILURE);
        }
        config->report = StringWriterData(contents);
        config->report_len = SafeStringLength(config->report);

        Log(LOG_LEVEL_NOTICE, "Read %d bytes for report contents", config->report_len);

        if (config->report_len == 0)
        {
            Log(LOG_LEVEL_ERR, "Report file contained no bytes");
            exit(EXIT_FAILURE);
        }

        Log(LOG_LEVEL_DEBUG, "Got report file contents: %s", config->report);
    }
```

Reply:
Yes, that looks like the best approach. :)